torch/distributed/rpc/rref_proxy.py

```diff
@@ -15,24 +15,66 @@ def _local_invoke_async_execution(rref, func_name, args, kwargs):
 def _invoke_rpc(rref, rpc_api, func_name, timeout, *args, **kwargs):
     # Since rref._get_type can potentially issue an RPC, it should respect the
     # passed in timeout here.
-    rref_type = rref._get_type(timeout=timeout)
-
-    _invoke_func = _local_invoke
-    # Bypass ScriptModules when checking for async function attribute.
-    bypass_type = issubclass(rref_type, torch.jit.ScriptModule) or issubclass(
-        rref_type, torch._C.ScriptModule
-    )
-    if not bypass_type:
-        func = getattr(rref_type, func_name)
-        if hasattr(func, "_wrapped_async_rpc_function"):
-            _invoke_func = _local_invoke_async_execution
-
-    return rpc_api(
-        rref.owner(),
-        _invoke_func,
-        args=(rref, func_name, args, kwargs),
-        timeout=timeout
-    )
+    # if rpc_sync, allow get_type to block, else, add invoke logic as a callback
+    block_on_type = rpc_api in [torch.distributed.rpc.rpc_sync, torch.distributed.rpc.remote]
+    rref_type = rref._get_type(timeout=timeout, blocking=block_on_type)
+    should_wait_on_rref_type = isinstance(rref_type, torch._C.Future)
```

Comment:
Can we always make the call to  _get_type nonblocking? That way the code is same for all APIs (rpc_sync, rpc_async and remote). The code is easier to read that way and we don't change any guarantees we provide to the user since the rpc_api call ensures we block appropriately when needed.

Reply:
The main use case was our internal training that uses rref.rpc_async extensively, although, I will test it with remote as well

Reply:
Updated diff to always make it non-blocking by default.